cpp/src/arrow/array/data.h

```diff
@@ -434,6 +435,23 @@ struct ARROW_EXPORT ArraySpan {
     return GetValues<T>(i, this->offset);
   }
 
+  // Access a buffer's data as a span
+  template <typename T>
+  util::span<const T> GetSpan(int i, int64_t length) const {
+    int buffer_length = static_cast<size_t>(buffers[i].size) / sizeof(T);
+    assert(length > 0 && length + offset <= buffer_length);
+    util::span<const T> data_span(buffers[i].data_as<T>(), buffer_length);
+    return data_span.subspan(offset, length);
```

Comment:
`GetSpan` should and will be inlined a lot, and the pre-conditions (that should be documented on the docstring) checked by the `assert` mean we don't need the extra truncation logic that `subspan` has.

`return util::span<const T>(buffers[i].data_as<T>() + this->offset, length);`